include/directx/d3dx12.h

```diff
@@ -4037,6 +4037,665 @@ class CD3DX12_NODE_MASK_SUBOBJECT
+        m_dRootSignature.HighestVersion = (D3D_ROOT_SIGNATURE_VERSION)0;
+        return S_OK;
+    }
+
+    // Helper funcion to decide the highest feature level
+    HRESULT QueryHighestFeatureLevel()
+    {
+        HRESULT result;
+
+        // Check against a list of all feature levels present in d3dcommon.h
+        // Needs to be updated for future feature levels
+        D3D_FEATURE_LEVEL allLevels[] = {
+            D3D_FEATURE_LEVEL_12_2,
+            D3D_FEATURE_LEVEL_12_1,
+            D3D_FEATURE_LEVEL_12_0,
+            D3D_FEATURE_LEVEL_11_1,
+            D3D_FEATURE_LEVEL_11_0,
+            D3D_FEATURE_LEVEL_10_1,
+            D3D_FEATURE_LEVEL_10_0,
+            D3D_FEATURE_LEVEL_9_3,
+            D3D_FEATURE_LEVEL_9_2,
+            D3D_FEATURE_LEVEL_9_1,
+            D3D_FEATURE_LEVEL_1_0_CORE
+        };
+
+        D3D12_FEATURE_DATA_FEATURE_LEVELS dFeatureLevel;
+        dFeatureLevel.NumFeatureLevels = sizeof(allLevels) / sizeof(D3D_FEATURE_LEVEL);
+        dFeatureLevel.pFeatureLevelsRequested = allLevels;
+
+        result = m_pDevice->CheckFeatureSupport(D3D12_FEATURE_FEATURE_LEVELS, &dFeatureLevel, sizeof(D3D12_FEATURE_DATA_FEATURE_LEVELS));
+        if (SUCCEEDED(result)) {
+            m_eMaxFeatureLevel = dFeatureLevel.MaxSupportedFeatureLevel;
+        } else {
+            m_eMaxFeatureLevel = (D3D_FEATURE_LEVEL)0;
+
+            if (result == DXGI_ERROR_UNSUPPORTED) { // Indicates that none are supported
+                result = S_OK;
+            }
+        }
+        return result;
+    }
+
+    // Helper function to initialize local protected resource session types structs
+    HRESULT QueryProtectedResourceSessionTypes(UINT NodeIndex, UINT Count)
+    {
+        auto& CurrentPRSTypes = m_dProtectedResourceSessionTypes[NodeIndex];
+        CurrentPRSTypes.NodeIndex = NodeIndex;
+        CurrentPRSTypes.Count = Count;
+        CurrentPRSTypes.TypeVec.resize(CurrentPRSTypes.Count);
+        CurrentPRSTypes.pTypes = CurrentPRSTypes.TypeVec.data();
+
+        return m_pDevice->CheckFeatureSupport(D3D12_FEATURE_PROTECTED_RESOURCE_SESSION_TYPES, &m_dProtectedResourceSessionTypes[NodeIndex], sizeof(D3D12_FEATURE_DATA_PROTECTED_RESOURCE_SESSION_TYPES));
+    }
+
+private: // Member data
+    // Pointer to the underlying device
+    ID3D12Device* m_pDevice;
+
+    // Stores the error code from initialization
+    HRESULT m_hStatus;
+
+    // Feature support data structs
+    D3D12_FEATURE_DATA_D3D12_OPTIONS m_dOptions;
+    D3D_FEATURE_LEVEL m_eMaxFeatureLevel;
+    D3D12_FEATURE_DATA_GPU_VIRTUAL_ADDRESS_SUPPORT m_dGPUVASupport;
+    D3D12_FEATURE_DATA_SHADER_MODEL m_dShaderModel;
+    D3D12_FEATURE_DATA_D3D12_OPTIONS1 m_dOptions1;
+    std::vector<D3D12_FEATURE_DATA_PROTECTED_RESOURCE_SESSION_SUPPORT> m_dProtectedResourceSessionSupport;
+    D3D12_FEATURE_DATA_ROOT_SIGNATURE m_dRootSignature;
+    std::vector<D3D12_FEATURE_DATA_ARCHITECTURE1> m_dArchitecture1;

```

Comment:
Continuation from last PR: std::vector is dynamic allocation which can cause memory failure during actual runs. Is it possible to replace these vectors (whose size is dependent on the number of graphics node in the machine) with statically allocated arrays, such as std::array?